File: python/examples/expense_tracker/expense_row/expense_row.py
"""Expense row component."""

import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))))

from runtime import UIApplication
from expense_row_ui import ExpenseRowUI


class ExpenseRow(ExpenseRowUI):
    """Interactive row used inside the expenses scroll view."""

    def configure(self, record, categories, on_delete=None, on_change=None):
        self.record = record
        self._on_delete = on_delete
        self._on_change = on_change
        self._categories = list(categories)
        self._sync_from_record()

    def on_paid_change(self, value):
        self.record["paid"] = bool(value)
        if self._on_change:
            self._on_change(self.record)

    def on_category_change(self, value):
        self.record["category"] = str(value)
        if self._on_change:
            self._on_change(self.record)

    def on_delete_btn_click(self):
        if self._on_delete:
            self._on_delete(self.record)

    def _sync_from_record(self):
        self.item.text = self.record.get("item", "")
        self.amount.text = f"{float(self.record.get('amount', 0)):.2f}"
        self.paid.value = bool(self.record.get("paid", False))
        self.category._options = list(self._categories)
        category = self.record.get("category") or (self._categories[0] if self._categories else "")
        self.category.selected_item = category if category in self._categories else (self._categories[0] if self._categories else "")


def main():
    app = UIApplication()
    window = ExpenseRow()
    window.configure({"item": "Coffee", "amount": 3.5, "category": "Food", "paid": True}, ["Food", "Other"])
    app.open(window)
    return app.run()


if __name__ == "__main__":
    sys.exit(main())

File: go/examples/expense_tracker/expense_row/expense_row.go
package expense_row

import (
	"fmt"

	uimd "uimd"
)

type DeleteCallback func(string)
type ChangeCallback func(string, string, bool)

type ExpenseRowView struct {
	ID       string
	Item     string
	Amount   float64
	Category string
	Paid     bool
}

type ExpenseRow struct {
	*ExpenseRowUI
	row        ExpenseRowView
	categories []string
	onDelete   DeleteCallback
	onChange   ChangeCallback
}

func NewExpenseRow(row ExpenseRowView, categories []string, onDelete DeleteCallback, onChange ChangeCallback) *ExpenseRow {
	view := &ExpenseRow{
		ExpenseRowUI: NewExpenseRowUI(),
		row:          row,
		categories:   append([]string{}, categories...),
		onDelete:     onDelete,
		onChange:     onChange,
	}
	view.SetEventHandler(view)
	view.syncFromRow()
	return view
}

func (row *ExpenseRow) ID() string {
	return row.row.ID
}

func (row *ExpenseRow) OnPaidChange(string) {
	row.row.Paid = row.Paid.Checked
	if row.onChange != nil {
		row.onChange(row.row.ID, selectedComboValue(row.Category), row.row.Paid)
	}
}

func (row *ExpenseRow) OnCategoryChange(value string) {
	row.row.Category = value
	if row.onChange != nil {
		row.onChange(row.row.ID, row.row.Category, row.Paid.Checked)
	}
}

func (row *ExpenseRow) OnDeleteBtnClick() {
	if row.onDelete != nil {
		row.onDelete(row.row.ID)
	}
}

func (row *ExpenseRow) syncFromRow() {
	row.Item.SetText(row.row.Item)
	row.Amount.SetText(fmt.Sprintf("%.2f", row.row.Amount))
	row.Paid.SetChecked(row.row.Paid)
	row.Category.Options = append([]string{}, row.categories...)
	selectComboValue(row.Category, row.row.Category)
}

func selectComboValue(comboBox *uimd.ComboBox, value string) {
	if comboBox == nil {
		return
	}
	for index, option := range comboBox.Options {
		if option == value {
			comboBox.SetSelectedIndex(index)
			return
		}
	}
	comboBox.SetSelectedIndex(0)
}

func selectedComboValue(comboBox *uimd.ComboBox) string {
	if comboBox == nil || comboBox.SelectedIndex < 0 || comboBox.SelectedIndex >= len(comboBox.Options) {
		return ""
	}
	return comboBox.Options[comboBox.SelectedIndex]
}
